File: src/Grimly.Core/Services/ClipboardService.cs
using System.Windows;
using System.Windows.Automation;
using Grimly.Native;

namespace Grimly.Services;

/// <summary>
/// Capture-to-internal / output-to-system workflow.
///
/// <para>
/// The app never writes captured or processed text to the system clipboard
/// during the capture and revision phases. All intermediate state lives in
/// the view-model's <c>WorkingText</c> (the internal buffer). The ONLY place
/// we write to the system clipboard is <see cref="PasteTextAsync"/>, which
/// runs exclusively when the user explicitly hits the Paste button — and at
/// that moment we both write to the clipboard and fire Ctrl+V so the target
/// app ingests the text.
/// </para>
/// </summary>
public interface IClipboardService
{
    /// <summary>
    /// Capture the target app's current text selection into an internal
    /// string — the caller stores it in the view-model's buffer. Does NOT
    /// write to, read from, or in any way disturb the system clipboard.
    ///
    /// <para>
    /// Works by reading the selection directly from the target app's
    /// accessibility tree (UIA <c>TextPattern</c>). Apps that don't expose
    /// UIA selection (some Chromium/Electron builds, custom text surfaces)
    /// return null — the caller then shows the user a friendly "can't read
    /// selection here" message.
    /// </para>
    /// </summary>
    Task<string?> GetSelectedTextAsync(IntPtr targetWindow);

    /// <summary>
    /// Export the internal buffer to the system clipboard and paste it into
    /// the target app. The ONLY clipboard-writing operation in this service.
    /// </summary>
    Task PasteTextAsync(string text, IntPtr targetWindow);
}

public sealed class ClipboardService : IClipboardService
{
    // ─────────────────────────────────────────────────────────────────────
    // CAPTURE PHASE — never touches the system clipboard
    // ─────────────────────────────────────────────────────────────────────

    public async Task<string?> GetSelectedTextAsync(IntPtr targetWindow)
    {
        // Make sure the target app has focus before we ask UIA about its
        // selection: AutomationElement.FocusedElement reports the system-wide
        // focused element, so we need the target to actually have it.
        var currentForeground = NativeMethods.GetForegroundWindow();
        if (targetWindow != IntPtr.Zero && targetWindow != currentForeground)
        {
            NativeMethods.SetForegroundWindow(targetWindow);
            await Task.Delay(80);
        }

        // UIA TextPattern runs on a background thread because the COM calls
        // can block for tens of ms. Returns the selected text as a plain
        // string — no clipboard involved, and no chance of leaving a "c" in
        // the user's document from a keystroke race.
        return await Task.Run(TryGetSelectedTextViaUIA);
    }

    /// <summary>
    /// Query the target app's accessibility tree for its current text
    /// selection. Returns null if UIA isn't usable for this focused element
    /// (no TextPattern) or the selection is empty.
    /// </summary>
    private static string? TryGetSelectedTextViaUIA()
    {
        try
        {
            var focused = AutomationElement.FocusedElement;
            if (focused == null) return null;

            if (!focused.TryGetCurrentPattern(TextPattern.Pattern, out var patternObj))
                return null;

            var textPattern = (TextPattern)patternObj;
            var selections = textPattern.GetSelection();
            if (selections == null || selections.Length == 0) return null;

            // -1 means "give me the whole range" — full text of the first selection.
            var text = selections[0].GetText(-1);
            return string.IsNullOrWhiteSpace(text) ? null : text;
        }
        catch
        {
            // AutomationElement can throw ElementNotAvailable, COM timeouts,
            // or unauthorized-access exceptions on apps with restricted UIA.
            return null;
        }
    }

    // ─────────────────────────────────────────────────────────────────────
    // OUTPUT PHASE — the only place we write to the system clipboard
    // ─────────────────────────────────────────────────────────────────────

    public async Task PasteTextAsync(string text, IntPtr targetWindow)
    {
        try { Clipboard.SetText(text); } catch { }

        if (targetWindow != IntPtr.Zero)
        {
            NativeMethods.SetForegroundWindow(targetWindow);
            await Task.Delay(100);
        }

        // WM_PASTE first — fires directly at the target's focused control,
        // no keyboard event queue, no modifier-timing races. Falls back to a
        // simulated Ctrl+V for apps that don't implement WM_PASTE (Chromium,
        // custom surfaces).
        bool pasted = false;
        if (targetWindow != IntPtr.Zero)
            pasted = NativeMethods.TrySendPasteMessage(targetWindow);

        if (!pasted)
            await NativeMethods.SendCtrlComboAsync(NativeMethods.VK_V);
    }
}

File: mac/Grimly/Services/ClipboardService.swift
import AppKit
import ApplicationServices
import Carbon.HIToolbox

/// Capture-to-internal / output-to-system workflow.
///
/// The app never writes captured or processed text to the system pasteboard
/// during the capture and revision phases. All intermediate state lives in
/// the view-model's `workingText` (the internal buffer). The ONLY place we
/// write to the pasteboard is `pasteText(_:previousApp:)`, which runs
/// exclusively when the user explicitly hits the Paste button — at that
/// moment we both write to the pasteboard and fire Cmd+V so the target app
/// ingests the text.
class ClipboardService {

    // MARK: - Capture phase — never touches the pasteboard

    /// Capture the target app's current text selection via the Accessibility
    /// API. Returns nil if AX can't extract a selection (target app doesn't
    /// expose accessibility, or nothing is selected).
    ///
    /// The old implementation sent Cmd+C, read the pasteboard, and restored
    /// the original pasteboard contents. That worked but briefly disturbed
    /// the user's clipboard. The AX path reads the selection directly from
    /// the target app's accessibility tree — no pasteboard involvement.
    func getSelectedText(previousApp: NSRunningApplication?) async -> String? {
        // The target needs to be frontmost for AX queries to reliably find
        // its focused element. Re-activate if we know which app to return to.
        if let app = previousApp {
            app.activate()
            try? await Task.sleep(nanoseconds: 100_000_000)
        }

        let pid: pid_t
        if let app = previousApp {
            pid = app.processIdentifier
        } else if let frontmost = NSWorkspace.shared.frontmostApplication {
            pid = frontmost.processIdentifier
        } else {
            return nil
        }

        return readSelectedText(forPID: pid)
    }

    /// Read `kAXSelectedTextAttribute` from the focused UI element of the
    /// given process. Safe to call from any thread; all the AX calls here are
    /// synchronous CF operations.
    private func readSelectedText(forPID pid: pid_t) -> String? {
        let appElement = AXUIElementCreateApplication(pid)

        var focusedValue: AnyObject?
        let focusResult = AXUIElementCopyAttributeValue(
            appElement,
            kAXFocusedUIElementAttribute as CFString,
            &focusedValue
        )
        guard focusResult == .success, let focusedElement = focusedValue else {
            return nil
        }

        // `focusedValue` comes back as AnyObject but is really an AXUIElement.
        // Unsafe bit-cast through CFTypeRef is the idiomatic way across this
        // API surface — AXUIElement isn't directly bridgeable.
        let focusedAX = focusedElement as! AXUIElement

        var selectedValue: AnyObject?
        let selResult = AXUIElementCopyAttributeValue(
            focusedAX,
            kAXSelectedTextAttribute as CFString,
            &selectedValue
        )
        guard selResult == .success, let text = selectedValue as? String else {
            return nil
        }

        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : text
    }

    // MARK: - Output phase — the only place we write to the pasteboard

    /// Write the internal buffer to the system pasteboard and paste it into
    /// the target app. The ONLY pasteboard-writing operation in this service.
    func pasteText(_ text: String, previousApp: NSRunningApplication?) async {
        let pasteboard = NSPasteboard.general
        pasteboard.clearContents()
        pasteboard.setString(text, forType: .string)

        if let app = previousApp {
            app.activate()
            try? await Task.sleep(nanoseconds: 100_000_000)
        }

        // Release any physically-held modifier keys before injecting Cmd+V so
        // the user's earlier hotkey doesn't taint the combo.
        releaseModifiers()
        sendKeyCombo(keyCode: UInt16(kVK_ANSI_V), flags: .maskCommand)
    }

    // MARK: - Keyboard helpers (used only by the Output phase)

    private func releaseModifiers() {
        let modifierKeys: [UInt16] = [
            UInt16(kVK_Command),
            UInt16(kVK_Option),
            UInt16(kVK_Shift),
            UInt16(kVK_Control),
        ]

        let source = CGEventSource(stateID: .combinedSessionState)

        for keyCode in modifierKeys {
            if let event = CGEvent(keyboardEventSource: source, virtualKey: keyCode, keyDown: false) {
                event.flags = []
                event.post(tap: .cghidEventTap)
            }
        }

        // Brief pause to let the key-up events register.
        usleep(50_000)
    }

    private func sendKeyCombo(keyCode: UInt16, flags: CGEventFlags) {
        let source = CGEventSource(stateID: .combinedSessionState)

        guard let keyDown = CGEvent(keyboardEventSource: source, virtualKey: keyCode, keyDown: true),
              let keyUp = CGEvent(keyboardEventSource: source, virtualKey: keyCode, keyDown: false) else {
            return
        }

        keyDown.flags = flags
        keyUp.flags = flags

        keyDown.post(tap: .cghidEventTap)
        keyUp.post(tap: .cghidEventTap)
    }
}
